### PR TITLE
HPCC-30458 Parquet plugin needs timestamp support

### PR DESCRIPTION
**Pull request by jackdelv**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
@jackdelv I'm not sure about the efficiency of going via text for most of the field types.  These look like fairly low level routines that will be called a lot.

**Comment by jackdelv** [comment]
@ghalliday I agree that it was inefficient to go through text for every type. Now it only happens if we are converting data stored in text to other types.

File: plugins/parquet/parquetembed.cpp [hunk 1]
```diff
@@ -838,6 +838,125 @@ int64_t ParquetRowBuilder::currArrayIndex()
...
+        case LargeStringType:
+            return (*array_visitor)->large_string_arr->GetView(currArrayIndex());
+        case DecimalType:
+            return (*array_visitor)->size == 128 ? (*array_visitor)->dec_arr->GetView(currArrayIndex()) : (*array_visitor)->large_dec_arr->GetView(currArrayIndex());
+        default:
+            failx("Unimplemented Parquet type for field with name %s.", field->name);
+    }
+}
+
+__int64 ParquetRowBuilder::getCurrValue(const RtlFieldInfo *field)
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:926]: possible getCurrIntValue() to make it clear what kind of value it is returning

File: plugins/parquet/parquetembed.cpp [hunk 2]
```diff
@@ -838,6 +838,125 @@ int64_t ParquetRowBuilder::currArrayIndex()
...
+{
+    switch ((*array_visitor)->type)
+    {
+        case BoolType:
+            return (*array_visitor)->bool_arr->Value(currArrayIndex());
+        case IntType:
+            return getSigned(array_visitor, currArrayIndex());
+        case UIntType:
+            return getUnsigned(array_visitor, currArrayIndex());
+        case DoubleType:
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:936]: Potential for a getReal() function similar to getSigned/unsigned which returned values for Double/HalfFloat/Float which is also called from getRealResult()

**Review by ghalliday — COMMENTED** [review]
A couple of minor comments, but looks close to ready.

File: plugins/parquet/parquetembed.cpp (hunk 2 above)
Comment (jackdelv) [reply plugins/parquet/parquetembed.cpp]: Added a getReal() function and removed extra types. Added getReal to getCurrIntValue() and getRealResult().

File: plugins/parquet/parquetembed.cpp (hunk 1 above)
Comment (jackdelv) [reply plugins/parquet/parquetembed.cpp]: Changed.

File: plugins/parquet/parquetembed.cpp [hunk 3]
```diff
@@ -872,6 +872,21 @@ unsigned __int64 getUnsigned(std::shared_ptr<ParquetArrayVisitor> *array_visitor
     }
 }
 
+double getReal(std::shared_ptr<ParquetArrayVisitor> *array_visitor, int index)
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:875]: Minor: These functions could be static (to avoid the function names being exported globally).  Don't change for this PR though.

File: plugins/parquet/parquetembed.cpp [hunk 4]
```diff
@@ -1015,14 +1020,10 @@ double ParquetRowBuilder::getRealResult(const RtlFieldInfo *field)
...
-            return (*array_visitor)->double_arr->Value(currArrayIndex());
-        case FloatType:
-            return (*array_visitor)->float_arr->Value(currArrayIndex());
-        default:
-            return getCurrValue(field);
-    }
+    if ((*array_visitor)->type == RealType)
+        return getReal(array_visitor, currArrayIndex());
+    else
+        return getCurrIntValue(field);
```
Comment (ghalliday) [inline plugins/parquet/parquetembed.cpp:1026]: Strings will not be converted to real values.  However I would be tempted to create a separate PR to fix that and merge this one as-is.

**Review by ghalliday — APPROVED** [review]
I like the further cleanup, switching to use the size.

File: plugins/parquet/parquetembed.hpp [hunk 5]
```diff
@@ -211,16 +216,65 @@ class ParquetArrayVisitor : public arrow::ArrayVisitor
...
+    {
+        date64_arr = &array;
+        type = DateType;
+        size = 64;
+        return arrow::Status::OK();
+    }
+    arrow::Status Visit(const arrow::TimestampArray &array)
+    {
+        timestamp_arr = &array;
+        type = TimestampType;
```
Comment (jakesmith) [inline plugins/parquet/parquetembed.hpp:236]: is it okay to leave size=0 ?

**Review by jakesmith — APPROVED** [review]
@jackdelv 1 question, but looks good.

File: plugins/parquet/parquetembed.hpp (hunk 5 above)
Comment (jackdelv) [reply plugins/parquet/parquetembed.hpp:236]: Yes, the size only gets used for the Signed, Unsigned, and Real types to differentiate between the different variations of each that can be stored in parquet.

Comment (jackdelv) [reply plugins/parquet/parquetembed.hpp:236]: Yes, the size only gets used for the Signed, Unsigned, and Real types to differentiate between the different variations of each that can be stored in parquet.

**Comment by ghalliday** [comment]
@jackdelv please squash
